[DROOLS-1629] avoid Mixing of ObjectMarshallingStrategies of the same class using…

ngs-mtech opened the pull request:
… NamedObjectMarshallingStrategy

Solving issue DROOLS-1629

Added more logging to improve evidence of the issue if its happening.

mariofusco commented:
ok to test

ngs-mtech commented:
I can´t see what have happened!

mariofusco commented:
Jenkins is reporting the following compilation error:

[ERROR] /home/jenkins/workspace/drools-pullrequests/drools-persistence/drools-persistence-jpa/src/main/java/org/drools/persistence/jpa/marshaller/JPAPlaceholderResolverStrategy.java:[46,8] org.drools.persistence.jpa.marshaller.JPAPlaceholderResolverStrategy is not abstract and does not override abstract method getName() in org.drools.core.marshalling.NamedObjectMarshallingStrategy

Will you fix it or should I give a look?

ngs-mtech commented:
Sorry Mario, here you have the fix.

Regards.

mariofusco commented:
We believe that what you need can be easily achieved without this pull request as demonstrated here https://gist.github.com/DuncanDoyle/f8cc93ba0559fefd8e1c64b1db43f643

I also tried the solution proposed by @DuncanDoyle and it works for me. Please let me know if we're overlooking something. If this also work for you I'll close this pull request and reject the corresponding jira.

ngs-mtech commented:
Hi Mario,
      
      I think I didnt explain the problem we have clear enough. We have
      a marshalling strategy (a common one for the whole collection of
      modules in our project), that implements some extra features that
      we found interesting regarding variable persistence.
      
      The problem comes from the fact that, when using a marshalling
      strategy in a kie module,  it gets registered in the list together
      with the rest (as expected), but when a marshalling strategy is
      employed in dependant modules, it gets registered multiple times
      (same class, different parameters). From this approach, while
      unmarshalling, the system cannot guess which is the correct
      strategy that was used for the marshalling.
      
      In the case you did understand this, suggesting a multiple class
      implementation with same code looks not only poor, but unpractical
      approach, as it would require a redeployment of the
      workbench/server war every time a new module needs to use that
      strategy. Remember that this class must be available at war
      classloader level (not module level).
      
      Taking into account that your solution (test update) is not
      solving the problem but hiding it, by making the test that makes
      it "visible" work, we do believe that the one we 'pulled' is, in
      our humble opinion, much more robust, flexible and practical.  
      
      Hope this lets the problem we found more clear and makes you
      understand that what you propose is not an acceptable solution,
      neither for our customers, neither for anyone in the community. 
      
      Best Regards,
      
      
      
      El 05/07/2017 a las 16:50, Mario Fusco escribió:
    
    
      We believe that what you need can be easily achieved without
        this pull request as demonstrated here https://gist.github.com/DuncanDoyle/f8cc93ba0559fefd8e1c64b1db43f643
      I also tried the solution proposed by @DuncanDoyle and it works for me.
        Please let me know if we're overlooking something. If this also
        work for you I'll close this pull request and reject the
        corresponding jira.
      —
        You are receiving this because you authored the thread.
        Reply to this email directly, view it on GitHub, or mute the thread.
      
        
          
          
        
        
      
      {"api_version":"1.0","publisher":{"api_key":"05dde50f1d1a384dd78767c55493e4bb","name":"GitHub"},"entity":{"external_key":"github/kiegroup/drools","title":"kiegroup/drools","subtitle":"GitHub repository","main_image_url":"https://cloud.githubusercontent.com/assets/143418/17495839/a5054eac-5d88-11e6-95fc-7290892c7bb5.png","avatar_image_url":"https://cloud.githubusercontent.com/assets/143418/15842166/7c72db34-2c0b-11e6-9aed-b52498112777.png","action":{"name":"Open in GitHub","url":"https://github.com/kiegroup/drools"}},"updates":{"snippets":[{"icon":"PERSON","message":"@mariofusco in #1354: We believe that what you need can be easily achieved without this pull request as demonstrated here https://gist.github.com/DuncanDoyle/f8cc93ba0559fefd8e1c64b1db43f643\r\n\r\nI also tried the solution proposed by @DuncanDoyle and it works for me. Please let me know if we're overlooking something. If this also work for you I'll close this pull request and reject the corresponding jira."}],"action":{"name":"View Pull Request","url":"https://github.com/kiegroup/drools/pull/1354#issuecomment-313126345"}}}

mswiderski commented:
unless I am missing something but what I think is the intended implementation of the strategies is to hide such complexity in the impl instead of making the framework to deal with it. What I am trying to say is that you should have this logic required to manage different modules. 

With that said I'd recommend to use single strategy that includes the logic to figure out what should actually be unmarshalled in what way. Similar implementation was done for case management and its case file marshalling that has single strategy but can handler different types of data - https://github.com/kiegroup/jbpm/blob/master/jbpm-case-mgmt/jbpm-case-mgmt-impl/src/main/java/org/jbpm/casemgmt/impl/marshalling/CaseFileInstanceMarshallingStrategy.java

ngs-mtech commented:
Hello Maciej
      
      I am able to agree with the design approach that you propose for
      Marshalling strategies but what we are talking about here is a bug
      [https://issues.jboss.org/browse/DROOLS-1629],
      a bug that says the marshalling strategy you use to marshall an
      object is different from the one used to unmarshall, as the
      marshalling task uses the method accept(Object ) to fetch the
      correct marshalling strategy and the deserialization uses the
      class name to fetch the target.
      
      This is a fact that is happening regardless of the design approach
      of the strategies you comment on, which, in fact, is the same as
      my chief architect proposed.
      
      It is important to emphasize, as you know, from another reported
      incidence, that KieWorkbench autoregisters the
      JPAPlaceholderMarshallingStrategy each time the persistence.xml is
      modified.
      
      When one module depends from another, the situation of having two
      ObjectMarshallinStrategies (JPAPlaceholderMarshallingStrategy)
      implemented with the same class is imposed by the environment.
      
      If you find a better solution for this error, we can apply it, but
      doing nothing is not a solution to a reproducible error.
      
      In fact the code indicated in the CaseManagement suffers from the
      same error.
      
      Regards
      
      
      
      El 05/07/2017 a las 20:57, Maciej Swiderski escribió:
    
    
      unless I am missing something but what I think is the intended
        implementation of the strategies is to hide such complexity in
        the impl instead of making the framework to deal with it. What I
        am trying to say is that you should have this logic required to
        manage different modules.
      With that said I'd recommend to use single strategy that
        includes the logic to figure out what should actually be
        unmarshalled in what way. Similar implementation was done for
        case management and its case file marshalling that has single
        strategy but can handler different types of data - https://github.com/kiegroup/jbpm/blob/master/jbpm-case-mgmt/jbpm-case-mgmt-impl/src/main/java/org/jbpm/casemgmt/impl/marshalling/CaseFileInstanceMarshallingStrategy.java
      —
        You are receiving this because you authored the thread.
        Reply to this email directly, view it on GitHub, or mute the thread.
      
        
          
          
        
        
      
      {"api_version":"1.0","publisher":{"api_key":"05dde50f1d1a384dd78767c55493e4bb","name":"GitHub"},"entity":{"external_key":"github/kiegroup/drools","title":"kiegroup/drools","subtitle":"GitHub repository","main_image_url":"https://cloud.githubusercontent.com/assets/143418/17495839/a5054eac-5d88-11e6-95fc-7290892c7bb5.png","avatar_image_url":"https://cloud.githubusercontent.com/assets/143418/15842166/7c72db34-2c0b-11e6-9aed-b52498112777.png","action":{"name":"Open in GitHub","url":"https://github.com/kiegroup/drools"}},"updates":{"snippets":[{"icon":"PERSON","message":"@mswiderski in #1354: unless I am missing something but what I think is the intended implementation of the strategies is to hide such complexity in the impl instead of making the framework to deal with it. What I am trying to say is that you should have this logic required to manage different modules. \r\n\r\nWith that said I'd recommend to use single strategy that includes the logic to figure out what should actually be unmarshalled in what way. Similar implementation was done for case management and its case file marshalling that has single strategy but can handler different types of data - https://github.com/kiegroup/jbpm/blob/master/jbpm-case-mgmt/jbpm-case-mgmt-impl/src/main/java/org/jbpm/casemgmt/impl/marshalling/CaseFileInstanceMarshallingStrategy.java"}],"action":{"name":"View Pull Request","url":"https://github.com/kiegroup/drools/pull/1354#issuecomment-313194958"}}}

mswiderski commented:
then I am really missing something big here...

> Avoid Mixing of ObjectMarshallingStrategies of the same class

is the title of this PR so to address this problem - use single strategy and not many of the same class.

With that said can we go back few steps and clearly define what's the issue. Could you please provide single test case that illustrates the bug?

ngs-mtech commented:
Hi Maciej,

We already have done a [unit test]( https://github.com/ngs-mtech/drools/blob/99e74e6bb98907661c9fe039b2ea5e50f371157a/drools-core/src/test/java/org/drools/core/marshalling/impl/ObjectMarshallingStrategyStoreTest.java) exposing the issue, using the same ObjectMarshallingStrategy twice in array, is the one that DuncanDoyle has changed to show the workarround "use single strategy and not many of the same class",  by declaring different subclasses hiding the real problem.

The problem is not a development affair, is a real situation imposed by the platform that occurs while using dependant modules having each one their own marshalling strategies defined.

It will take me some time to create a kie workbench repository demostrating the issue while using a **dependant module**.

In this scenario, both modules will have their custom persistence unit and a declaration in their deployment descriptor of their own JPAPlaceholderResolverStrategy( "<module>", classLoader) defined. 

Later, in the runtime environment of the dependant module, all the marshalling strategies (the ones from base module and the ones from the dependant module) are created and registered in the execution environment, so after that there are instances of the same class but from different modules registered. 

![multiple-objectmarshallingstrategies-on-context-with-same-class](https://user-images.githubusercontent.com/5487398/27922798-8f12b6da-627c-11e7-8380-69d5f2d11c1b.png)


In this scenario, while doing the marshalling/unmarshalling of session variables, depending on the index of the ObjectMarshallingStrategy in the environment, the task of marshalling or unmarshalling will work or fail.

Regards

mswiderski commented:
So is the double marshaller in the runtime a side effect of using defendant kjars that have same marshaller already defined? 
Assuming you register them via deployment descriptor then there a option to define how to deal with merging of multiple descriptors where one of them is to override all so only the top level will be used.

ngs-mtech commented:
Yes Maciej, this is another way to skip the bug, but it is not a solution.

I can't undestand why we need to apply a workaround having a pull request with a working solution that solves the issue. We have implemented the solution taking into account legacy implementations, so is not a rupturistic approach.

Where is the failure?

mswiderski commented:
and how is this protecting the case where you have two kjars that declare the same strategy (with same name). They will be still registered, isn't it? So we might end up with exact same issue...

in case they are slightly differently configured then the issue is still there, right? Or the assumption is that then the name should take into account the config that could differ to distinguish it...

mswiderski commented on drools-core/src/main/java/org/drools/core/marshalling/NamedObjectMarshallingStrategy.java:
```diff
@@ -0,0 +1,35 @@
+	 * or targets (Different persistence units, Different Document Content storages, etc ).
+	 * @return The unique name in the project environment for the marshalling strategy represented
+	 */
+	public String getName();
```
I don't think there is a need for new interface, we should make use of Java 8 features and extend the org.kie.api.marshalling.ObjectMarshallingStrategy with new method getName that is a default method which returns class.getName().

ngs-mtech replied:
Changing an interface is very hard to support legacy implementations but your proposal of default implementation of the getName method in interface is a nice approach to the solution. 
The problem is that the ObjectMarshallingStrategy belongs to kie-api project that is outside drools project. This means a different pull request, don't it?

mswiderski commented on drools-core/src/main/java/org/drools/core/marshalling/impl/IdentityPlaceholderResolverStrategy.java:
```diff
@@ -113,8 +118,11 @@ public Context createContext() {
             objects.put(idEntry.getValue(), idEntry.getKey());
+            nameIds.append( idEntry.getKey() ).append( "," );
         }
+        this.name = IdentityPlaceholderResolverStrategy.class.getName() + nameIds.toString();
```
wouldn't this be a breaking compatibility change in case ids where given? If so then we should not add it here and then let users extend it if needed

ngs-mtech replied:
I think the behavior will not change unless you add new ones during the execution, right? Is this the case?

mswiderski commented on drools-core/src/main/java/org/drools/core/marshalling/impl/ObjectMarshallingStrategyStoreImpl.java:
```diff
@@ -16,15 +16,38 @@
+        		name = ((NamedObjectMarshallingStrategy)strategy).getName();
+        	}else{
+        		name = strategy.getClass().getName();
+        	}
```
this if/else is not needed when we use default method on org.kie.api.marshalling.ObjectMarshallingStrategy

mswiderski commented on drools-core/src/main/java/org/drools/core/marshalling/impl/ObjectMarshallingStrategyStoreImpl.java:
```diff
@@ -16,15 +16,38 @@
+        		name = strategy.getClass().getName();
+        	}
+        	if( names.contains( name ) ){
+        		logger.warn( "Multiple ObjectMarshallingStrategies with the same name:" + name + " strange behaviour could occurr");
```
if we want to add this then we should fail if there are duplicates instead of logging warning that strange behaviour can occur as that is not really helping and can lead to issues you faced and might not be easy to recover

mswiderski commented on drools-core/src/main/java/org/drools/core/marshalling/impl/ObjectMarshallingStrategyStoreImpl.java:
```diff
@@ -60,9 +83,22 @@ public ObjectMarshallingStrategy getStrategyObject(String strategyClassName) {
-           if( strategiesList[i].getClass().getName().equals(strategyClassName) ) {
-               return strategiesList[i];
-           }
+        	if( strategiesList[i] instanceof NamedObjectMarshallingStrategy 
```
same comment here, there is no need for this block anymore once we have default getName on org.kie.api.marshalling.ObjectMarshallingStrategy, s we should always check strategy.getName

mswiderski commented on drools-core/src/main/java/org/drools/core/marshalling/impl/PersisterHelper.java:
```diff
@@ -239,9 +241,13 @@ public static void writeRuntimeDefinedClasses( MarshallerWriteContext context,
+			Builder _strat = ProtobufMessages.Header.StrategyIndex.newBuilder()
                                      .setId( entry.getValue().intValue() )
-                                     .setName( entry.getKey().getClass().getName() );
+                                     .setName( name );
```
same here, it should only be entry.getKey().getName()

mswiderski commented on drools-core/src/main/java/org/drools/core/marshalling/impl/SerializablePlaceholderResolverStrategy.java:
```diff
@@ -22,21 +22,26 @@
 
     private int index;
-    
+    private String name = SerializablePlaceholderResolverStrategy.class.getName();
```
not needed when there is default method on org.kie.api.marshalling.ObjectMarshallingStrategy

mswiderski commented on drools-core/src/test/java/org/drools/core/marshalling/impl/NamedCustomMarshallingStrategy.java:
```diff
@@ -0,0 +1,22 @@
+package org.drools.core.marshalling.impl;
```
missing license header

mswiderski commented on drools-core/src/test/java/org/drools/core/marshalling/impl/ObjectMarshallingStrategyStoreTest.java:
```diff
@@ -0,0 +1,208 @@
+package org.drools.core.marshalling.impl;
```
missing license header
